### PR TITLE
[SYCL][Fusion] Enable -Werror for fusion library

### PR DESCRIPTION
**Pull request by jopperm**
Treat warnings as errors, and fix the last remaining warnings in the fusion library.

### CODE REVIEW

File: sycl-fusion/CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,9 @@ set(SYCL_JIT_BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 # directories, similar to how clang/CMakeLists.txt does it.
 set(LLVM_SPIRV_INCLUDE_DIRS "${LLVM_MAIN_SRC_DIR}/../llvm-spirv/include")
 
+# Set library-wide warning options.
+set(SYCL_FUSION_WARNING_FLAGS -Werror -Wall -Wextra)
```
Comment (sommerlukas) [inline sycl-fusion/CMakeLists.txt:13]: I'm ok with having this on by default (in particular for CI), but there should be some CMake option to turn it off. `-Werror` can be extremely annoying while debugging an issue in code or when your testing WIP changes.

File: sycl-fusion/common/CMakeLists.txt [hunk 2]
```diff
@@ -1,16 +1,25 @@
-add_llvm_library(sycl-fusion-common
+add_library(sycl-fusion-common
```
Comment (sommerlukas) [inline sycl-fusion/common/CMakeLists.txt:1]: Can we add a comment for future reference to explain why we're taking the manual approach here?

File: sycl-fusion/jit-compiler/CMakeLists.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 
-add_llvm_library(sycl-fusion
+add_library(sycl-fusion
```
Comment (sommerlukas) [inline sycl-fusion/jit-compiler/CMakeLists.txt:2]: Same as above, a comment would be good.

File: sycl-fusion/jit-compiler/CMakeLists.txt [hunk 4]
```diff
@@ -29,14 +29,20 @@ add_llvm_library(sycl-fusion
    ${LLVM_TARGETS_TO_BUILD}
 )
 
+target_compile_options(sycl-fusion PRIVATE -fno-rtti ${SYCL_FUSION_WARNING_FLAGS})
```
Comment (sommerlukas) [inline sycl-fusion/jit-compiler/CMakeLists.txt:32]: Should we really set the `-fno-rtti` flag unconditionally here?

File: sycl-fusion/CMakeLists.txt (hunk 1 above)
Comment (jopperm) [reply sycl-fusion/CMakeLists.txt]: Makes sense, added!

**Comment by jopperm** [comment]
I rolled back the change from `add_llvm_library` to `add_library` in my last commit. LLVM's helper contains some logic to determine whether RTTI should be enabled or not, and I think duplicating that in our code would negate any benefit of following the recommended `add_library`-way for out-of-tree libraries. @victor-eds is that ok?

File: sycl-fusion/passes/target/TargetFusionInfo.cpp [hunk 5]
```diff
@@ -408,8 +408,8 @@ class SPIRVTargetFusionInfo : public TargetFusionInfoImpl {
   }
 
   Function *createRemapperFunction(
-      const Remapper &R, BuiltinKind K, StringRef OrigName, Module *M,
-      const jit_compiler::NDRange &SrcNDRange,
+      const Remapper &R, BuiltinKind K, StringRef OrigName [[maybe_unused]],
```
Comment (victor-eds) [inline sycl-fusion/passes/target/TargetFusionInfo.cpp:411]:
```suggestion
      const Remapper &R, BuiltinKind K, [[maybe_unused]] StringRef,
```

Really just a minor, but I think having the attribute before the type and dropping the argument name is more readable.

File: sycl-fusion/passes/target/TargetFusionInfo.cpp [hunk 6]
```diff
@@ -551,7 +551,7 @@ class NVPTXAMDGCNTargetFusionInfoBase : public TargetFusionInfoImpl {
                                                   uint32_t Idx) const = 0;
 
   Value *getGlobalIDWithoutOffset(IRBuilderBase &Builder,
-                                  const NDRange &FusedNDRange,
+                                  const NDRange &FusedNDRange [[maybe_unused]],
```
Comment (victor-eds) [inline sycl-fusion/passes/target/TargetFusionInfo.cpp:554]: Same

File: sycl-fusion/passes/target/TargetFusionInfo.cpp [hunk 7]
```diff
@@ -687,8 +687,9 @@ class NVPTXAMDGCNTargetFusionInfoBase : public TargetFusionInfoImpl {
     switch (K) {
     case BuiltinKind::NumWorkGroupsRemapper:
     case BuiltinKind::GroupIDRemapper:
-      return WrapValInFunc(
-          [&](uint32_t Idx) { return Builder.getInt32(R.getDefaultValue(K)); });
+      return WrapValInFunc([&](uint32_t Idx [[maybe_unused]]) {
```
Comment (victor-eds) [inline sycl-fusion/passes/target/TargetFusionInfo.cpp:690]: Same

File: sycl-fusion/CMakeLists.txt [hunk 8]
```diff
@@ -9,6 +9,14 @@ set(SYCL_JIT_BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 # directories, similar to how clang/CMakeLists.txt does it.
 set(LLVM_SPIRV_INCLUDE_DIRS "${LLVM_MAIN_SRC_DIR}/../llvm-spirv/include")
 
+# Set library-wide warning options.
```
Comment (victor-eds) [inline sycl-fusion/CMakeLists.txt:12]:
```suggestion
separate_arguments(LLVM_DEFINITIONS_LIST NATIVE_COMMAND ${LLVM_DEFINITIONS})

# Set library-wide warning options.
```

And `target_compile_definitions` to all targets we define.

**Review by victor-eds — CHANGES_REQUESTED** [review]
Maybe conditionally adding `-fno-exceptions` and `-fno-rtti` (add by default) is not a bad idea. We'd have to add some logic as `SYCL_FUSION_ENABLE_EH` would set `SYCL_FUSION_ENABLE_RTTI` (`message(AUTHOR_WARNING ...`).

File: sycl-fusion/passes/target/TargetFusionInfo.cpp (hunk 5 above)
Comment (jopperm) [reply sycl-fusion/passes/target/TargetFusionInfo.cpp]: Ah yes, thanks!

**Comment by victor-eds** [comment]
> I rolled back the change from `add_llvm_library` to `add_library` in my last commit. LLVM's helper contains some logic to determine whether RTTI should be enabled or not, and I think duplicating that in our code would negate any benefit of following the recommended `add_library`-way for out-of-tree libraries. @victor-eds is that ok?

My issue with that is we'd be adding our libraries to `llvm-libraries` target... Can we keep the `add_library` approach and use `llvm_update_compile_flags` instead?

**Review by sommerlukas — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Comment by jopperm** [comment]
We've discussed offline that we keep the use `add_llvm_library` for the sake of this PR, and consider switching to `add_library` in a future PR.

**Comment by sommerlukas** [comment]
Post-commit failure: https://github.com/intel/llvm/pull/12830

Fix for the post-commit failure is here: https://github.com/intel/llvm/pull/13013

**Pull request merged**